### DISCUSSION TITLE
Extend or customize syntax highlighting

### DISCUSSION THREAD

**Discussion by EDKarlsson** [eclipse-langium/langium#604] [Q&A] [answered]
I only started using this yesterday so I may have not stumbled on it. Is there a way to customize or extend the generated syntax highlighting? I'd prefer to not have everything go under "keyword.control". I know I could add semantic highlighting. do a embedded language or injection. I tried injection but I'm new to using that feature of vscode extensions. My use case is having 2 different styles of comments.

Comment belongs to the type it is being referred to while a note should be ignored.
Note starts with `//*` and comment has only 1 `/` 
```ANTLR
terminal COMMENT: /\/\*[\s\S]*\*\//
hidden terminal NOTE: /\/\/*[\s\S]*?\*\//;
```

This results in anything within `/* .... */` not having any highlighting, well it does have highlighting that follows the patterns of the language and but no specific rule in the "repository". I tried adding a return to the `COMMENT` terminal but didn't do anything. It would be nice to be able to define the scope of a rule within the grammar.

**Comment by msujew** [comment, marked as answer]
Hi @EDKarlsson, the generated textMate syntax highlighting file is basically just a starting help for new grammars. Anyone who wants more fine-grained control over their syntactical highlighting should deactivate the `textMate` option in their `langium-config.json` and write/modify their textMate file by hand. It's basically impossible to infer accurate syntactical highlighting just based on a grammar. That's also what we did for the [Langium syntax highlighting](https://github.com/langium/langium/blob/main/packages/langium-vscode/data/langium.tmLanguage.json).

Regarding your comments: I'm pretty sure that this style of commenting will lead to a lot of problems down the line. Your `COMMENT` terminal contains a greedy `[\s\S]*` regex (compare to the non-greedy `[\s\S]*?` regex), which basically parses/lexes as "read everything in the file until the end of time itself". Have you confirmed that this doesn't introduce any parsing errors? Additionally, it will definitely override the `NOTE` terminal, since they are ambiguous, so that one will never be lexed. I would recommend you to remove the `COMMENT` terminal. Just use the `NOTE` terminal and call the `findCommentNode` function to find comments for a CST node (you can use `astNode.$cstNode` to get the CST node of an AST node). 

> It would be nice to be able to define the scope of a rule within the grammar.

I think this is (\*drum roll\*) out of scope for Langium. Our grammar language is already complicated enough as it is, and syntax highlighting can be easily done manually, so I don't really see a reason to include this into the grammar language.

**Reply by AviCraimer** [reply]
Hi,

You mentioned disabling the textMate generation, but you didn't specify how to do this using langium-config.json settings. I cannot find documentation for those settings on the Langium website. At the Langium language link you sent it appears that you are not using langium:generate at all (at least there is no script for it in your package.json). Please let me know if there is a way to use langium:generate without generating the textMate grammar. Thank you.

**Reply by msujew** [reply]
Hey @AviCraimer, you just need to remove the `textMate` entry in the `langium-config.json`. Here an example without a textmate entry:

https://github.com/eclipse-langium/langium/blob/d2eda289fd7283d375292b7a77fc4e0abaf64707/packages/langium/langium-config.json#L3-L12

> At the Langium language link you sent it appears that you are not using langium:generate at all

We do, it's just a different package that does the generate call:

https://github.com/eclipse-langium/langium/blob/d2eda289fd7283d375292b7a77fc4e0abaf64707/packages/langium/package.json#L62

**Reply by AviCraimer** [reply]
Thank you!

**Comment by EDKarlsson** [comment]
That works as well. I couldn't find any documentation on deactivating the tmLanguage generation. Thanks for answering my question!

I did try it and it does not overwrite the NOTE terminal. Unfortunately I do not have control of the grammar. This is a fun side project implementing SysML v2. I got decently far with my first attempt using textX (https://github.com/textX/textX) but it doesn't have any concept of fragments and rules started becoming unwieldy just to avoid creating an element representation of a rule

**Reply by EDKarlsson** [reply]
Here is an example of it working. I deactivated the tmLanguage generation which is a perfectly acceptable solution. There isn't documentation about doing so being okay. Though I will probably have to generate it until I've finished with the entire language spec since it would be a hassle to constantly have to add each rule to the the syntax grammar.

![image](https://user-images.githubusercontent.com/7072370/179794330-672460fa-b361-4c77-829b-d7dfca815a54.png)

**Reply by EDKarlsson** [reply]
Found an easy workaround to keep auto generated syntax highlighting and having a custom one by using https://www.npmjs.com/package/json-merger and adding a build step. I have Langium write to `sysml.generated.tmlLanguage.json` then use this `template.tmLanguage.json` and finally write it to `sysml.tmLanguage.json`

```json
{
    "$merge": {
        "source": {
            "$import": "syntaxes/sysml.generated.tmLanguage.json"
        },
        "with": {
            "patterns": {
                    "$concat": [{
                         "include": "#sysml-comments"
                    }
                ],
            },
            "repository": {
                "sysml-comments": {
                    "patterns": [
                        {
                            "name": "string.unquoted.sysml",
                            "begin": "/\\*",
                            "beginCaptures": {
                                "0": {
                                    "name": "punctuation.definition.string.begin.sysml"
                                }
                            },
                            "end": "\\*/",
                            "endCaptures": {
                                "0": {
                                    "name": "punctuation.definition.string.end.sysml"
                                }
                            }
                        }
                    ]
                }
            }
        }
    }
}

```

**Reply by msujew** [reply]
> Here is an example of it working.

Oh I see now, I didn't notice that the `NOTE` and `COMMENT` terminals started differently `/*` and `//*`. Yeah, now it all makes sense.

> Found an easy workaround to keep auto generated syntax highlighting and having a custom one

Cool, that's good to know. We might want to include that in the documentation on textMate grammars that's supposed to come eventually.

**Comment by ponyatov** [comment]
@msujew Can I highlight language elements based on some primitive semantic analysis?

For example, can I use different colors for different type constants (type-dependent coloring):
```evento
const TGRAM = '@dponyatov';
const TAB = 0x09`u8;
const PI = 3.1415926535;
```

or differently highlight names created by `const`, `let` and `mut`:
```evento
const TGRAM = '@dponyatov';
let dialect = 'FORTH';
mut compile = false;
```

**Reply by cdietrich** [reply]
This would be done by implementing semantic highlighting as eg in https://github.com/eclipse-langium/langium/blob/f53d8c7f756d1364f6078aea302f8144a627fc3e/packages/langium/src/grammar/lsp/grammar-semantic-tokens.ts#L8
